lib/socket.c

```diff
@@ -653,14 +657,35 @@ iscsi_read_from_socket(struct iscsi_context *iscsi)
 		in->data_pos += count;
 	}
 
+	/* not enough data */
 	if (in->data_pos < data_size) {
 		return 0;
 	}
 
 	ISCSI_LIST_ADD_END(&iscsi->inqueue, in);
 	iscsi->incoming = NULL;
 
+	*pdu = in;
+	return 1;
+}
 
+// TODO(sahlberg) change this to 1 for AROS which doesn't support NONBLOCK io
+// actually
+#define READ_PDU_BATCH_COUNT 32
+
+static int iscsi_read_from_socket(struct iscsi_context *iscsi) {
+	int i;
+	for (i = 0; i < READ_PDU_BATCH_COUNT; ++i) {
+		struct iscsi_in_pdu *pdu;
+		int ret = iscsi_read_pdu_from_socket(iscsi, &pdu);
+		/* NODATA */
+		if (ret == 0) {
+			break;
+		}
+		if (ret < 0) {
+			return -1;
+		}
```

Comment:
I still don't think call ioctl(FIONREAD) is a good idea because currently 100G nic does 12.5G/s throughput which means 12.5 kb per us. you must miss many pdu during several syscalls after ioctl. for the case of 10G nic, you miss 1.25 kilobytes per us.